### PR TITLE
Add instrumentation and lightweight tracing to Speedy CEK machine

### PR DESCRIPTION
**Pull request by nickchapman-da**
Add instrumentation and lightweight tracing to Speedy execution.
- Disabled by default.
- Enabled with compile time flags
- When these flags are constant `false`, there should be no cost to to the code

I have benchmarked this branch against master, and can see no measurable performance impact. If there is any, it is not visible in the noise.

Flags to enable are:
-  `enableInstrumentation` 
-  `enableLightweightStepTracing`

Instrumentation:
- tracks #pushes and maximum-depth for the `kont`-stack and the `env`ironment
- counts #steps
- classifies  each step:
    - for `CtrlExpr`, what kind of expression is it?
    - for `CtrlValue`, what kind of continuation is at the head of the `kont`-stack?
- the information is printed when the machine finalizes

Lightweight tracing shows an abbreviated output line per step:
- the control expression / value
- the size of the environment
- the depth of the `kont`-stack

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 1]
```diff
@@ -132,6 +132,11 @@ private[lf] final case class Compiler(packages: PackageId PartialFunction Packag
   def unsafeCompile(expr: Expr): SExpr =
     validate(closureConvert(Map.empty, 0, translate(expr)))
 
+  @throws[PackageNotFound]
+  @throws[CompilationError]
+  def unsafeCompile(sexpr: SExpr): SExpr =
+    validate(closureConvert(Map.empty, 0, sexpr))
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:138]: I am not sure we should use "Compile" for the name of this function.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/PrettyLightweight.scala [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+  def pp(prim: Prim): String = prim match {
+    case PBuiltin(b) => s"$b"
+    case PClosure(expr, fvs) =>
+      s"clo[${commas(fvs.map(pp))}]:${pp(expr)}"
+  }
+
+  def commas(xs: Seq[String]): String = xs.mkString(",")
+
+  final case class UnknownV(v: SValue) extends RuntimeException(v.toString)
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/PrettyLightweight.scala:92]:
```suggestion
  final case class UnknownValue(v: SValue) extends RuntimeException(v.toString)
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 3]
```diff
@@ -53,12 +57,56 @@ object Speedy {
       var localContracts: Map[V.ContractId, (Ref.TypeConName, SValue)],
       // global contract discriminators, that are discriminators from contract created in previous transactions
       var globalDiscriminators: Set[crypto.Hash],
+      /* Used when enableLightweightStepTracing is true */
+      var steps: Int,
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:61]: We could wrapped this variable and the following ones in a new object (call something like statics) to limit the extension this endless list of parameter. 
You will pay the small cost of indirection only when instrumentation is running.

File: daml-lf/interpreter/perf/BUILD.bazel [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+load(
+    "//bazel_tools:scala.bzl",
+    "da_scala_binary",
+)
+
+da_scala_binary(
+    name = "explore",
+    srcs = glob(["src/main/**/*.scala"]),
+    main_class = "com.daml.lf.speedy.explore.Explore",
+    visibility = ["//visibility:public"],
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/perf/BUILD.bazel:13]: Why making it public ?

**Comment by nickchapman-da** [comment]
@remy Thank you for your suggestions, which have been addressed.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 5]
```diff
@@ -59,19 +96,53 @@ object Speedy {
...
+    /* env manipulation... */
+
+    @inline def envDepth(): Int = env.size()
+
     @inline def getEnv(i: Int): SValue = env.get(env.size - i)
-    @inline def popEnv(count: Int): Unit =
+
+    @inline def pushEnv(v: SValue): Unit = {
+      env.add(v)
+      if (enableInstrumentation) {
```
Comment (hurryabit) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:129]: Are we sure we don't have to pay for this even when `enableInstrumentation` is false? Would it make sense to subclass `Speedy` and override these methods? Or do we have to pay for additional dynamic dispatch then?

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/PrettyLightweight.scala [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+  def ppKont(k: Kont): String = k match {
+    case KPop(n) => s"KPop($n)"
+    case KArg(es) => s"KArg(${commas(es.map(pp))})"
+    case KFun(prim, extendedArgs, arity) =>
+      s"KFun(${pp(prim)}/$arity,[${commas(extendedArgs.asScala.map(pp))}])"
+    case KPushTo(_, e) => s"KPushTo(_, ${pp(e)})"
+    case KCacheVal(_, _) => "KCacheVal"
+    case KLocation(_) => "KLocation"
+    case KMatch(_) => "KMatch"
+    case KCatch(_, _, _) => "KCatch" //never seen
```
Comment (hurryabit) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/PrettyLightweight.scala:46]: You need to use `submitMustFail` to get these.

**Pull request merged**